lib/utils/process-jobs.ts

```diff
@@ -149,18 +159,21 @@ export const processJobs = function(this: Agenda, extraJob: Job) {
    * @returns {undefined}
    */
   async function jobQueueFilling(name: string) {
-    // Don't lock because of a limit we have set (lockLimit, etc)
-    if (!shouldLock(name)) {
-      debug('lock limit reached in queue filling for [%s]', name);
-      return;
-    }
+    debug('jobQueueFilling: %s isJobQueueFilling: %s', name, self._isJobQueueFilling.has(name));
+    self._isJobQueueFilling.set(name, true);
 
-    // Set the date of the next time we are going to run _processEvery function
-    const now = new Date();
-    self._nextScanAt = new Date(now.valueOf() + self._processEvery);
-
-    // For this job name, find the next job to run and lock it!
     try {
+      // Don't lock because of a limit we have set (lockLimit, etc)
+      if (!shouldLock(name)) {
+        debug('lock limit reached in queue filling for [%s]', name);
+        return; // goes to finally block
+      }
+
+      // Set the date of the next time we are going to run _processEvery function
+      const now = new Date();
+      self._nextScanAt = new Date(now.valueOf() + self._processEvery);
+
+      // For this job name, find the next job to run and lock it!
       const job = await self._findAndLockNextJob(name, definitions[name]);
```

Comment:
if I compare your changes to mine (please also feel free to do this, and bring changes back to the main branch :)). I have found another check for "shoudLock" I added after the "findAndLockNextJob". https://github.com/hokify/agenda/blob/e7ea1deb69e55f71a1f6defdfb69773940689eb8/src/JobProcessor.ts#L331
 The reason for this is that it can happen that findAndLockNextJob takes a while to finish (e.g. datebase is slow, or we just have a lot of new jobs coming in right now), then it can happen that another jobqueufilling (or locking on the fly) can fill up the queue in the meantime, just while we were waiting for "findandlocknextjob". therefore checking the locking conditions after it one more time, is a rare, but still possible case.

I just remember that this last issue is even more relevant in the main agenda version as it has one other issue regarding "findandlocknextjob" (see https://github.com/agenda/agenda/issues/1130). don't nail me on this, but as far as I remember this lead to jobs that are filling up the "lockedJobs", but are not executed in this order, which causes some behaviour that e.g. one job will fill up over the limit, and another one does not even get one "slot" (see also https://github.com/agenda/agenda/pull/1123#issuecomment-708342692). Or even better would be to bring in the test case and see if this is related.

Reply:
This 2nd issue is even more relevant, as the main agenda version has one other issue regarding "findandlocknextjob" (see https://github.com/agenda/agenda/issues/1130). don't nail me on this, but as far as I remember this lead to jobs that are filling up the "lockedJobs", but are not executed in this order, which causes some behaviour that e.g. one job will fill up over the limit, and another one does not even get one "slot".

Reply:
I agree that `_findAndLockNextJob()` can potentially be slow and cause troubles.

Reply:
@simllll I retested it with the https://github.com/agenda/agenda/pull/1119 merged.
the test still failed intermittently

I added the changes of this PR and the test worked.